### PR TITLE
[ios, macos] Add notes to APIs that rely on style and source ids

### PR DESCRIPTION
**Pull request by boundsj**
Fixes https://github.com/mapbox/mapbox-gl-native/issues/6616 for iOS and macOS

cc @1ec5

### CODE REVIEW

File: platform/darwin/src/MGLStyle.h [hunk 1]
```diff
@@ -177,6 +177,13 @@ static const NSInteger MGLStyleDefaultVersion = 9;
 /**
  Returns a layer that conforms to `MGLStyleLayer` if any layer with the given
  identifier was found.
+ 
+ @note Layer identifiers are subject to change and are not guaranteed to exist
```
Comment (1ec5) [inline platform/darwin/src/MGLStyle.h]: This is an awkward place to put “subject to change”. Maybe in the next sentence, say that the default style’s layers are subject to change.

File: platform/darwin/src/MGLStyle.h [hunk 2]
```diff
@@ -177,6 +177,13 @@ static const NSInteger MGLStyleDefaultVersion = 9;
...
+    accross styles or different versions of the same style. Applications that
+    use this API must set the style URL to an explicitly versioned style
+    using a convenience method like +[MGLStyle outdoorsStyleURLWithVersion:],
```
Comment (1ec5) [inline platform/darwin/src/MGLStyle.h]: Wrap this selector in backticks to enable autolinking in jazzy.

File: platform/darwin/src/MGLStyle.h [hunk 3]
```diff
@@ -177,6 +177,13 @@ static const NSInteger MGLStyleDefaultVersion = 9;
...
+    the “Style URL” IBInspectable in `MGLMapView`, or a manually constructed
```
Comment (1ec5) [inline platform/darwin/src/MGLStyle.h]: > `MGLMapView`’s “Style URL” inspectable in Interface Builder

**Comment by 1ec5** [comment]
Additional APIs to warn about:

* [ ] `-[MGLStyle removeLayer:]`
* [ ] `-[MGLStyle removeSource:]`
* [ ] `-[MGLStyle removeStyleClass:]`
* [ ] `-[MGLStyle removeImageForName:]`

**Comment by boundsj** [comment]
Before I move on to https://github.com/mapbox/mapbox-gl-native/pull/6886#issuecomment-258022837 I took another pass at the copy. Please let me know what you think.

File: platform/darwin/src/MGLStyle.h [hunk 4]
```diff
@@ -177,6 +177,14 @@ static const NSInteger MGLStyleDefaultVersion = 9;
 /**
  Returns a layer that conforms to `MGLStyleLayer` if any layer with the given
  identifier was found.
+ 
+ @note Layer identifiers are not guaranteed to exist across styles or different
+    versions of the same style. Applications that use this API must set the style
```
Comment (1ec5) [inline platform/darwin/src/MGLStyle.h]: s/must/must first/

I think that could be a point of confusion for developers.

File: platform/darwin/src/MGLStyle.h [hunk 5]
```diff
@@ -177,6 +177,14 @@ static const NSInteger MGLStyleDefaultVersion = 9;
 /**
  Returns a layer that conforms to `MGLStyleLayer` if any layer with the given
  identifier was found.
+ 
+ @note Layer identifiers are not guaranteed to exist across styles or different
+    versions of the same style. Applications that use this API must set the style 
+    URL to an explicitly versioned style using a convenience method like 
+    `+[MGLStyle outdoorsStyleURLWithVersion:]`, `MGLMapView`'s “Style URL” 
+    inspectable in Interface Builder, or a manually constructed NSURL. This
```
Comment (1ec5) [inline platform/darwin/src/MGLStyle.h]: Nit: put `NSURL` in backticks (for consistency with SDK classes that would get autolinked, even if this one doesn’t get linked).

File: platform/ios/src/MGLMapView.h [hunk 6]
```diff
@@ -1079,6 +1079,20 @@ IB_DESIGNABLE
...
+ purposes. It also has a set of convenience methods for creating style URLs of 
+ default styles provided by Mapbox.
+ 
+ @note The default styles provided by Mapbox contain sources and layers with
+    identifiers that will change over time. Applications that use APIs that 
+    manipulate a style's sources and layers must first set the style URL to an 
+    explicitly versioned style using a convenience method like
+    `+[MGLStyle outdoorsStyleURLWithVersion:]`, `MGLMapView`'s “Style URL”
+    inspectable in Interface Builder, or a manually constructed `NSURL`.
+ */
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.h]: Please update the macOS implementation of MGLMapView with these changes as well.

File: platform/darwin/src/MGLStyle.h [hunk 7]
```diff
@@ -226,6 +259,14 @@ static const NSInteger MGLStyleDefaultVersion = 9;
 /**
  Removes a source from the map view.
 
+ @note Source identifiers are not guaranteed to exist across styles or different
+    versions of the same style. Applications that use this API must first set the
+    style URL to an explicitly versioned style using a convenience method like
+    `+[MGLStyle outdoorsStyleURLWithVersion:]`, `MGLMapView`'s “Style URL”
+    inspectable in Interface Builder, or a manually constructed `NSURL`. This
+    approach also avoids layer identifer name changes that will occur in the default
+    style’s layers over time.
```
Comment (1ec5) [inline platform/darwin/src/MGLStyle.h]: s/layer/source/ (×2)

**Comment by boundsj** [comment]
@1ec5 what do you think about this now?

**Comment by 1ec5** [comment]
I think there’s still a typo in a couple places: https://github.com/mapbox/mapbox-gl-native/pull/6886#discussion_r87697081

**Comment by boundsj** [comment]
Corrected and rebased @1ec5

**Pull request merged**